mne/preprocessing/tests/test_regress.py

```diff
@@ -34,5 +35,90 @@ def test_regress_artifact():
-        regress_artifact(epochs, picks='eog', picks_artifact='eog')
+    # Regressing channels onto themselves should work
+    epochs, betas = regress_artifact(epochs, picks='eog', picks_artifact='eog')
+    assert np.ptp(epochs.get_data('eog')) < 1E-15  # constant value
+    assert_allclose(betas, 1)
+
+    # Test applying TAA.
+    raw_no_taa, betas = regress_artifact(raw, taa=False)
+    raw_taa, _ = regress_artifact(raw, taa=True, betas=betas)
+    # TAA should have increased signal amplitude
+    assert np.ptp(raw_no_taa.get_data('eeg')) < np.ptp(raw_taa.get_data('eeg'))
+
+
+@testing.requires_testing_data
+def test_eog_regression():
+    """Test regressing artifact data using the EOGRegression class."""
+    raw = read_raw_fif(raw_fname).pick(['eeg', 'eog', 'stim'])
+
+    # Test error regarding preloading
+    with pytest.raises(RuntimeError, match='requires raw data to be loaded'):
+        model = EOGRegression().fit(raw)
+
+    raw.load_data()
+
+    # Test regression on raw data
+    model = EOGRegression().fit(raw)
+    assert model.coef_.shape == (59, 1)  # 59 EEG channels, 1 EOG channel
+    raw_clean = model.apply(raw)
+    # Some signal must have been removed
+    assert np.ptp(raw_clean.get_data('eeg')) < np.ptp(raw.get_data('eeg'))
+
+    # Test regression on epochs
+    epochs = create_eog_epochs(raw)
+    model = EOGRegression().fit(epochs)
+    epochs = model.apply(epochs)
+    # Since these were blinks, they should be mostly gone
+    assert np.ptp(epochs.get_data('eeg')) < 1E-4
+
+    # Test regression on evoked
+    evoked = epochs.average('all')
+    model = EOGRegression().fit(evoked)
+    evoked = model.apply(evoked)
+    assert model.coef_.shape == (59, 1)
+    # Since this was a blink evoked, signal should be mostly gone
+    assert np.ptp(evoked.get_data('eeg')) < 1E-4
+
+    # Test regression on evoked and applying to raw, with different ordering of
+    # channels
+    raw_ = raw.copy().drop_channels(['EEG 001'])
+    raw_ = raw_.add_channels([raw.copy().pick(['EEG 001'])])
+    model = EOGRegression().fit(evoked)
+    model.apply(raw_)
+
+    # Test applying TAA.
+    raw_no_taa = EOGRegression().fit(raw).apply(raw, taa=False)
+    raw_taa = EOGRegression().fit(raw).apply(raw, taa=True)
+    # TAA should have increased signal amplitude
+    assert np.ptp(raw_no_taa.get_data('eeg')) < np.ptp(raw_taa.get_data('eeg'))
+
+    # Test in-place operation
+    raw_ = model.apply(raw, copy=False)
+    assert raw_ is raw
+    assert raw_._data is raw._data
+    raw_ = model.apply(raw, copy=True)
+    assert raw_ is not raw
+    assert raw_._data is not raw._data
+
+    # Test plotting with one channel type
+    fig = model.plot()
+    assert len(fig.axes) == 2  # (one topomap and one colorbar)
+    assert fig.axes[0].title.get_text() == 'eeg/EOG 061'
+
+    # Test plotting with multiple channel types
+    raw = read_raw_fif(raw_fname).pick(['meg', 'eeg', 'eog'])
```

Comment:
*nitpick* ideally we'd avoid loading the same data from disk twice in one test. Can this be refactored to load once at the beginning and make a copy where needed?  Or perhaps do the MEG+EEG test lines first, and then (using the same object) pick just the EEG chans and do the remaining tests?

Reply:
I'm sure it can be refactored that way, yes.